File: conversion.c
#include <stdio.h>
#include <math.h>
#include <string.h>

// ---------- Binary to Decimal ----------
int binaryToDecimal(long long bin) {
    int dec = 0, i = 0, rem;
    while (bin != 0) {
        rem = bin % 10;
        dec += rem * pow(2, i);
        bin /= 10;
        i++;
    }
    return dec;
}

// ---------- Octal to Decimal ----------
int octalToDecimal(int oct) {
    int dec = 0, i = 0;
    while (oct != 0) {
        dec += (oct % 10) * pow(8, i);
        oct /= 10;
        i++;
    }
    return dec;
}

// ---------- Hexadecimal to Decimal ----------
int hexToDecimal(char hex[]) {
    int len = strlen(hex);
    int dec = 0, base = 1;

    for (int i = len - 1; i >= 0; i--) {
        if (hex[i] >= '0' && hex[i] <= '9')
            dec += (hex[i] - '0') * base;
        else if (hex[i] >= 'A' && hex[i] <= 'F')
            dec += (hex[i] - 'A' + 10) * base;
        else if (hex[i] >= 'a' && hex[i] <= 'f')
            dec += (hex[i] - 'a' + 10) * base;

        base *= 16;
    }
    return dec;
}

// ---------- Decimal to Binary ----------
void decimalToBinary(int dec) {
    int bin[32], i = 0;
    while (dec > 0) {
        bin[i++] = dec % 2;
        dec /= 2;
    }
    for (int j = i - 1; j >= 0; j--)
        printf("%d", bin[j]);
}

// ---------- Decimal to Octal ----------
void decimalToOctal(int dec) {
    int oct[32], i = 0;
    while (dec > 0) {
        oct[i++] = dec % 8;
        dec /= 8;
    }
    for (int j = i - 1; j >= 0; j--)
        printf("%d", oct[j]);
}

// ---------- Decimal to Hex ----------
void decimalToHex(int dec) {
    char hex[32];
    int i = 0, rem;

    while (dec > 0) {
        rem = dec % 16;
        if (rem < 10)
            hex[i++] = rem + '0';
        else
            hex[i++] = rem - 10 + 'A';
        dec /= 16;
    }
    for (int j = i - 1; j >= 0; j--)
        printf("%c", hex[j]);
}

// ---------- MAIN MENU ----------
int main() {
    int choice, dec, oct;
    long long bin;
    char hex[20];

    printf("===== Number Base Conversion Menu =====\n");
    printf("1. Binary\n");
    printf("2. Decimal\n");
    printf("3. Octal\n");
    printf("4. Hexadecimal\n");
    printf("Choose source base: ");
    scanf("%d", &choice);

    int target;
    printf("Choose target base (1-Bin, 2-Dec, 3-Oct, 4-Hex): ");
    scanf("%d", &target);

    if (choice == 1) { // Binary
        printf("Enter binary number: ");
        scanf("%lld", &bin);
        dec = binaryToDecimal(bin);
    }
    else if (choice == 2) { // Decimal
        printf("Enter decimal number: ");
        scanf("%d", &dec);
    }
    else if (choice == 3) { // Octal
        printf("Enter octal number: ");
        scanf("%d", &oct);
        dec = octalToDecimal(oct);
    }
    else if (choice == 4) { // Hex
        printf("Enter hexadecimal number: ");
        scanf("%s", hex);
        dec = hexToDecimal(hex);
    }
    else {
        printf("Invalid source base!\n");
        return 0;
    }

    printf("Result: ");
    if (target == 1)
        decimalToBinary(dec);
    else if (target == 2)
        printf("%d", dec);
    else if (target == 3)
        decimalToOctal(dec);
    else if (target == 4)
        decimalToHex(dec);
    else
        printf("Invalid target base!");

    printf("\n");
    return 0;
}

File: conversion.py
# ---------- To Decimal ----------
def binary_to_decimal(b):
    dec = 0
    power = 0
    for digit in reversed(b):
        dec += int(digit) * (2 ** power)
        power += 1
    return dec


def octal_to_decimal(o):
    dec = 0
    power = 0
    for digit in reversed(o):
        dec += int(digit) * (8 ** power)
        power += 1
    return dec


def hex_to_decimal(h):
    dec = 0
    power = 0
    h = h.upper()

    for ch in reversed(h):
        if ch.isdigit():
            value = int(ch)
        else:
            value = ord(ch) - ord('A') + 10
        dec += value * (16 ** power)
        power += 1

    return dec


# ---------- From Decimal ----------
def decimal_to_binary(n):
    if n == 0:
        return "0"
    result = ""
    while n > 0:
        result = str(n % 2) + result
        n //= 2
    return result


def decimal_to_octal(n):
    if n == 0:
        return "0"
    result = ""
    while n > 0:
        result = str(n % 8) + result
        n //= 8
    return result


def decimal_to_hex(n):
    if n == 0:
        return "0"
    digits = "0123456789ABCDEF"
    result = ""
    while n > 0:
        result = digits[n % 16] + result
        n //= 16
    return result


# ---------- MAIN MENU ----------
print("===== Base Conversion Menu =====")
print("1. Binary")
print("2. Decimal")
print("3. Octal")
print("4. Hexadecimal")

source = int(input("Choose source base: "))
target = int(input("Choose target base (1-Bin, 2-Dec, 3-Oct, 4-Hex): "))
number = input("Enter number: ")

# Convert source to decimal
if source == 1:
    decimal_value = binary_to_decimal(number)
elif source == 2:
    decimal_value = int(number)
elif source == 3:
    decimal_value = octal_to_decimal(number)
elif source == 4:
    decimal_value = hex_to_decimal(number)
else:
    print("Invalid source base!")
    exit()

# Convert decimal to target
print("Result: ", end="")

if target == 1:
    print(decimal_to_binary(decimal_value))
elif target == 2:
    print(decimal_value)
elif target == 3:
    print(decimal_to_octal(decimal_value))
elif target == 4:
    print(decimal_to_hex(decimal_value))
else:
    print("Invalid target base!")
